pkg/audit/manager.go

```diff
@@ -253,7 +253,7 @@ func (am *Manager) audit(ctx context.Context) error {
 
 		err := am.addAuditResponsesToUpdateLists(updateLists, res, totalViolationsPerConstraint, totalViolationsPerEnforcementAction, timestamp)
 		if errs != nil {
-			return err
+			am.log.Error(err, "Auditing")
```

Comment:
Why are we squashing errors here and below?

Reply:
Before the pub-sub change `addAuditResponsesToUpdateLists` didn't really generate any error. By returning "Publishing" `err` and not "Audit" errors at line 256 and below, we are disrupting the audit, which results in `constraintStatus` not being updated with violations.

I am also open to ideas on handling publishing errors in other ways.

Reply:
But we are not squashing pubsub errors, we are squashing any error returned by `addAuditResponsesToUpdateLists()`, which seems much broader in scope than what you're intending (merely swallowing pubsub errors).

Reply:
Also, anticipating another comment:

What is our contract with users with regards to failing to publish an audit message?

The way this is written, it appears we are giving up on first failure, even if there is something like a transient networking issue.

What is our contract with users? When will events fail to be published?

Reply:
> But we are not squashing pubsub errors, we are squashing any error returned by addAuditResponsesToUpdateLists(), which seems much broader in scope than what you're intending (merely swallowing pubsub errors).

before pub-sub change, `addAuditResponsesToUpdateLists` didn't generate any error at all, it returned `nil`. So any errors returned now are only errors encountered in publishing the message. 🤔

Reply:
> What is our contract with users? When will events fail to be published?

Do we want to have a retry mechanism? since audit is periodic I assume those transient errors might get resolved by the next audit run and user eventually gets the audit violations. 🤔

Reply:
Maybe? But is a user expecting *some* of the audit results or *all* of the audit results per audit run to be sent to them.

To be clear, *all* is a theoretical impossibility, but there is a big difference between "drop on first error" and "have some amount of retry w/exponential backoff and jitter"

Reply:
@ritazh @sozercan would be helpful to have your voice as to user expectations here.

Reply:
Few chains of thought - 
1. Popular pubsub tools might already have a retry mechanism in place. Can we assume that when a driver fails to publish a message, the underlying provider/tool has exhausted the retries? 
2. Implementing retries on top of tool's retries might cause overhead for the messages that are not getting delivered because of legit errors.

Reply:
Sorry for the delay in looking at this.

@maxsmythe i think @JaydipGabani's point is `addAuditResponsesToUpdateLists` never returned an err before pubsub (it had the return type as an `error`, which is a bug I believe - let me know if I am missing something) https://github.com/open-policy-agent/gatekeeper/blob/v3.10.0/pkg/audit/manager.go#L693
So previously (pre pubsub), audit was fault tolerant and we always published to constraints

however, now that pubsub is actually returning an error as part of `addAuditResponsesToUpdateLists` this would have unintended consequences of actually returning an error and skipping the constraint update.
I think @JaydipGabani's change is intending to restore the previous behavior so we still update constraints even if there is an error.

Reply:
I think at some point we stopped returning errors b/c it was poorly-behaved. In general, audit code could probably use a refactor at some point. The core point is that we probably shouldn't be swallowing all errors unless it is part of some principled design for audit (which, again, we should do at some point).

If all we are doing is avoiding rocking the boat, then we should swallow pubsub failures close to the source. Anything grander is getting into redesign territory, and requires an understanding of audit as a whole to write/review.

Regardless, the larger point of this discussion thread is: what are the user expectations with regard to failed pubsub requests?

Reply:
IMO, the user expectation should be we return as many violations as possible even when we encounter errors. The audit process should be fault tolerant in that we log the errors, but an error should not impact all violations.